core/src/main/java/io/questdb/cairo/TxReader.java

```diff
@@ -0,0 +1,295 @@
+    protected long txn;
+    protected int symbolsCount;
+    protected long dataVersion;
+    protected long structureVersion;
+    protected long fixedRowCount;
+    protected long transientRowCount;
+    protected int partitionBy = -1;
+    protected long partitionTableVersion;
+    private VirtualMemory roTxMem;
+    private ReadOnlyMemory readOnlyTxMem;
+    private Timestamps.TimestampFloorMethod timestampFloorMethod;
+
+    public TxReader(FilesFacade ff, Path path) {
+        this.ff = ff;
+        this.path = new Path(path.length() + 10);
+        this.path.put(path);
+        this.rootLen = path.length();
+    }
+
+    public boolean attachedPartitionsContains(long ts) {
+        return findAttachedPartitionIndex(ts) > -1;
+    }
+
+    @Override
+    public void close() {
+        roTxMem = Misc.free(roTxMem);
+        path = Misc.free(path);
+    }
+
+    public int getPartitionsCount() {
+        return attachedPartitions.size() / LONGS_PER_TX_ATTACHED_PARTITION;
+    }
+
+    public long getDataVersion() {
+        return dataVersion;
+    }
+
+    public long getFixedRowCount() {
+        return fixedRowCount;
+    }
+
+    public long getMaxTimestamp() {
+        return maxTimestamp;
+    }
+
+    public long getMinTimestamp() {
+        return minTimestamp;
+    }
+
+    public long getPartitionSize(int i) {
+        return attachedPartitions.getQuick(i * LONGS_PER_TX_ATTACHED_PARTITION + PARTITION_SIZE_OFFSET);
+    }
+
+    public long getPartitionSizeByPartitionTimestamp(long ts) {
+        final int index = findAttachedPartitionIndex(getPartitionLo(ts));
+        if (index > -1) {
+            return attachedPartitions.getQuick(index + PARTITION_SIZE_OFFSET);
+        }
+        return -1;
+    }
+
+    public long getPartitionTableVersion() {
+        return partitionTableVersion;
+    }
+
+    public long getPartitionTimestamp(int i) {
+        return attachedPartitions.getQuick(i * LONGS_PER_TX_ATTACHED_PARTITION + PARTITION_TS_OFFSET);
+    }
+
+    public long getRowCount() {
+        return transientRowCount + fixedRowCount;
+    }
+
+    public long getStructureVersion() {
+        return structureVersion;
+    }
+
+    public long getTransientRowCount() {
+        return transientRowCount;
+    }
+
+    public long getTxn() {
+        return txn;
+    }
+
+    public long getTxEofOffset() {
+        return getTxMemSize(symbolsCount, attachedPartitions.size());
+    }
+
+    public void initPartitionBy(int partitionBy) {
+        this.timestampFloorMethod = partitionBy != PartitionBy.NONE ? getPartitionFloor(partitionBy) : null;
+        this.partitionBy = partitionBy;
+    }
+
+    public void open() {
+        assert this.roTxMem == null;
+        roTxMem = openTxnFile(ff, path, rootLen);
+        if (roTxMem instanceof ReadOnlyMemory) {
+            // In readonly mode Tx file has to call grow sometimes
+            this.readOnlyTxMem = (ReadOnlyMemory) roTxMem;
+        }
+    }
+
+    public void read() {
```

Comment:
there is API inconsistency here. This read is "unsafe", i.e. TableReader does ordered reads (sync) before calling into this method. However on TxWriter side `commit()` takes care of ordered writes so that TableWriter does not have to.

Reply:
This is unsafe read. I don't want to move save read out of TableReader it looks like it's a better place to me.
I can rename it to uncheckedRead()

Reply:
👍